---
title: "correction method 2"
author: "Marcelo Mora"
date: "16 de abril de 2017"
output: html_document
---
  
#Objective
  
  The objective of this analysis is correct the uptake samples for Rd contamination

This analysis is described in the following [link](http://rrresearch.fieldofscience.com/2017/04/a-new-plan-for-contamination-correction.html)


##load the working directory

First I need to set the working directory

```{r setup, include=FALSE}
# Set the path of the working directory; 
knitr::opts_knit$set(root.dir = 'C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/')

```


##Next I load the samples.

```{r eval = FALSE}

mix_depth<- read.csv("./correccion/partial3.csv", header = FALSE) #load a version of Josh csv file with only the MIX13, RR733 and UP13 columns. See below to read the code used to make this file 

#Note: the mix_depth file was editted from the original in the unix shell by extracting columns and rows we were interested (RR722, MIX13, UP13 aligned to NP only)

#awk -F, '{OFS=",";print $1,$2,$3,$6,$7,$9}' mix_data.csv > partial_mix.csv

#awk -F"," -v OFS="," '$1 == "npnn" { print $1,$2,$3,$4,$5,$6 }' partial_mix.csv > partial2.csv    

#awk -F"," -v OFS="," '$3 == "npnn" { print $1,$2,$3,$4,$5,$6 }' partial2.csv > partial3.csv    



#load the file in case of re-running
mix_depth_all<- read.csv(file = "C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/mix_depth_all.csv")



```


First step is loading all packages as well as the data. The data consists in all NP input samples (UP13, UP15), RR722 (Rd) and a MIX13(UP13 where 10% of the reads were replaced by Rd reads) aligned against Rd and NP concatenated chromosomes, as well as aligned against NP only. This has also been done with all the reads and with reads of aligning quality > 0.  


```{r, message= FALSE, warning = FALSE, eval = FALSE}

#############################load packages##############################
library(ggplot2)
#########################################################################



colnames(mix_depth)<- c("ref","filt","chr","pos","MIX13","RR722","UP13") # add headers

mix_depth_all <- mix_depth[grep("all", mix_depth$filt), 1:7] #get only rows from the alignment done from non-filtered reads

str(mix_depth) #check

```

#Normalize the reads

```{r eval = FALSE}


mapped_MIX13<- 2723028 #number of mapped reads

mapped_RR722<- 4065268 #number of mapped reads

mapped_UP13<- 2723792 #number of mapped reads

#calculate coverage per 1000 reads
mix_depth_all$MIX13_n<- (mix_depth_all$MIX13 * 1e+6)/mapped_MIX13
mix_depth_all$RR722_n<- (mix_depth_all$RR722 * 1e+6)/mapped_RR722
mix_depth_all$UP13_n<- (mix_depth_all$UP13 * 1e+6)/mapped_UP13

str(mix_depth_all) #check


View(mix_depth_all)

```


##Correction

The idea of the next analysis is that the percentage of contamination based on the concatenated alignment gives us an average contamination read amount, but does not gives us a position specific contamination amount. 

The following analysis allows us to estimate this per-position contamination amount by multipling the average contamination by position specific amount of Rd reads aligning to NP


```{r eval = FALSE}

#apply the correction (MIX13 - (RR722 * 0.1), where 0.1 is the 10% contamination we added to the sample. This percentage will have to be replaced in our uptake samples by the percentage of reads aligning to Rd in the concatenated alignment




mix_depth_all$MIX13_c<- mix_depth_all$MIX13_n - (0.1 * mix_depth_all$RR722_n)

str(mix_depth_all)

```

##Normalize the data for changes made by the correction

```{r eval = FALSE}

#Sum all the coverage numbers in the original (pre-correction) set (all 1.83 million of them, value 'A') and in the corrected set (value 'B'). Then normalize the corrected set of coverages by multiplying each value by the ratio of 'A' to 'B'.


n<- (sum(mix_depth_all$MIX13_n))/(sum(mix_depth_all$MIX13_c)) #get a ratio  of the sum of coverage from MIX13 without correction by MIX13 after correction

mix_depth_all$MIX13_nc<- mix_depth_all$MIX13_c * n   #multiply corrected MIX13 y the ratio

str(mix_depth_all)

mix_depth_all$MIX13_ratio<- mix_depth_all$MIX13_nc/mix_depth_all$UP13_n  #add this to the dataframe

flag<-  mix_depth_all$UP13_n[1:length(mix_depth_all$UP13_n)]>=10  #make the flag of low coverage positions in UP13

mix_depth_all$flag<- flag


#save the file
write.csv(mix_depth_all, file = "C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/mix_depth_all.csv", quote= FALSE)


```

Plot the ratio of normalized depth of corrected MIX13/UP13

```{r eval = FALSE}


p <- ggplot(aes(x = pos,y = log2(MIX13_ratio)), data = mix_depth_all) +
  geom_point(shape = 20, size = 1, aes(colour = flag))+
  scale_color_manual(values = c("TRUE" = "blue", "FALSE" = "red" ), name = "Input coverage",labels = c("TRUE" = "positions with more then 10 reads input", "FALSE" = "positions with less then 10 reads input")) +
  scale_x_continuous(breaks = seq(0 , 1914490, 500000), expand = c(0, 0))+
  scale_y_continuous(limits = c(-2, 2))+
  labs(x = "donor genome positions", y = "log2(ratio of normalized depth)") +
  ggtitle(" ratio of normalized depth of MIX13/UP13 ") +
  theme(legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=16),
        axis.title = element_text(size = 16, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/MIX13_UP13_n.png", width = 60, height = 30, units = "cm")

mix_depth_all$pos

p <- ggplot() +
  geom_point(aes(x = pos, y = MIX13_c), color = "blue",shape = 20, size = 1, data = test) +
  geom_point(aes(x = pos, y = UP13), color = "red",shape = 20, size = 1, data = test) +
  theme(
    panel.grid.minor = element_line(colour="white", size=0.5),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    axis.text  = element_text(size=16),
    axis.title = element_text(size = 16, face = "bold"))  
ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/test.png", width = 60, height = 30, units = "cm")


```