---
title: "correction method 2"
author: "Marcelo Mora"
date: "16 de abril de 2017"
output: html_document
---
  
#Objective
  
  The objective of this analysis is correct the uptake samples for Rd contamination

This analysis is described in the following [link](http://rrresearch.fieldofscience.com/2017/04/a-new-plan-for-contamination-correction.html)


##load the working directory

First I need to set the working directory

```{r setup, include=FALSE}
# Set the path of the working directory; 
knitr::opts_knit$set(root.dir = '~/DNA_uptake/')

```


##Next I load the samples.

```{r eval = FALSE}

library(data.table)

depth.rdnp.map<- fread("~/DNA_uptake/datasets/tables/rdnp.map.csv") #load concataneted samples aligned to concatenated rdnp genomes and only including reads with quality > 0  

str(depth.rdnp.map)

unique(depth.rdnp.map$V2)

depth.rdnp.all<- fread("~/DNA_uptake/datasets/tables/rdnp.all.f.csv") #load concataneted samples aligned to concatenated rdnp genomes including all reads 
 
unique(depth.rdnp.all$V2)


```


First step is loading all packages as well as the data. The data consists in all NP input samples (UP13, UP15), RR722 (Rd) and a MIX13(UP13 where 10% of the reads were replaced by Rd reads) aligned against Rd and NP concatenated chromosomes, as well as aligned against NP only. This has also been done with all the reads and with reads of aligning quality > 0.  


```{r, message= FALSE, warning = FALSE, eval = FALSE}

#############################load packages##############################
library(ggplot2)
library(dplyr)
#########################################################################

colnames(depth.rdnp.map)<- c("name","genome","pos","depth") # add headers
colnames(depth.rdnp.all)<- c("name","genome","pos","depth") # add headers


UP07.depth.rdnp.map<- depth.rdnp.map[grep("UP07", depth.rdnp.map$name), 1:4]
UP07.depth.rdnp.all<- depth.rdnp.all[grep("UP07", depth.rdnp.all$name), 1:4]

UP07.depth.rdnp<-data.frame(genome = UP07.depth.rdnp.map$genome, pos = UP07.depth.rdnp.map$pos, depth.map = UP07.depth.rdnp.map$depth, depth.all = UP07.depth.rdnp.all$depth)  



write.csv(UP07.depth.rdnp, "~/DNA_uptake/datasets/tables/UP07.depth.rdnp.csv")




```


Plot the ratio of normalized depth of corrected MIX13/UP13

```{r eval = FALSE}

UP07.depth.rdnp$ratio<- UP07.depth.rdnp$depth.all/(UP07.depth.rdnp$depth.map + 1)


rd<- dplyr::filter(UP07.depth.rdnp, genome == "KW20")

theme_set(theme_grey())

p <- ggplot(aes(x = pos,y = ratio), data = rd) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 1831585, 100000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 500))+
  labs(x = "KW20 genome positions", y = "ratio of all coverage divided by map+1") +
  ggtitle(" ratio of coverage of UP07  all/map+1") +
  theme(legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/UP07_2.png", width = 80, height = 40, units = "cm")


np<- dplyr::filter(UP07.depth.rdnp, genome == "NP")

theme_set(theme_grey())

p <- ggplot(aes(x = pos,y = ratio), data = np) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 1914386, 100000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 500))+
  labs(x = "NP genome positions", y = "ratio of all coverage divided by map+1") +
  ggtitle(" ratio of coverage of UP07  all/map+1") +
  theme(legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/UP07_np.png", width = 80, height = 40, units = "cm")



```

```{r}

UP15.depth.rdnp.map<- depth.rdnp.map[grep("UP15", depth.rdnp.map$name), 1:4]
UP15.depth.rdnp.all<- depth.rdnp.all[grep("UP15", depth.rdnp.all$name), 1:4]

UP15.depth.rdnp<-data.frame(genome = UP15.depth.rdnp.map$genome, pos = UP15.depth.rdnp.map$pos, depth.map = UP15.depth.rdnp.map$depth, depth.all = UP15.depth.rdnp.all$depth)  


UP15.depth.rdnp$ratio<- UP15.depth.rdnp$depth.all/(UP15.depth.rdnp$depth.map + 1)


rd<- dplyr::filter(UP15.depth.rdnp, genome == "KW20")

np<- dplyr::filter(UP15.depth.rdnp, genome == "NP")

theme_set(theme_grey())

p <- ggplot(aes(x = pos,y = ratio), data = rd) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 1831585, 100000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 500))+
  labs(x = "KW20 genome positions", y = "ratio of all coverage divided by map+1") +
  ggtitle(" ratio of coverage of UP15  all/map+1") +
  theme(legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/UP15.png", width = 80, height = 40, units = "cm")



ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/UP15_raw_all.png", width = 80, height = 40, units = "cm")

slice(rd,1:10000) %>% ggplot(aes(x = pos,y = depth.all)) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 10000, 1000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "KW20 genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to KW20, all reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_all_section_rd.png", width = 60, height = 30, units = "cm")

slice(np,1:10000) %>% ggplot(aes(x = pos,y = depth.all)) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 10000, 1000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "NP genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to NP, all reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_all_section_np.png", width = 60, height = 30, units = "cm")




slice(rd,1:10000) %>% ggplot(aes(x = pos,y = depth.map)) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 10000, 1000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "KW20 genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to KW20, map > 0 reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_map_section_rd.png", width = 60, height = 30, units = "cm")

slice(np,1:10000) %>% ggplot(aes(x = pos,y = depth.map)) +
  geom_point(shape = 20, size = 1)+
  scale_x_continuous(breaks = seq(0 , 10000, 1000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "NP genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to NP, map > 0 reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_map_section_np.png", width = 60, height = 30, units = "cm")


ggplot(aes(x = pos,y = depth.map), data = rd) +
  geom_point(shape = ".", size = 1)+
  scale_x_continuous(breaks = seq(0 , 1831585, 100000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "KW20 genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to KW20, map > 0 reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_map_section_rd_whole.png", width = 80, height = 30, units = "cm")

ggplot(aes(x = pos,y = depth.map), data = np) +
  geom_point(shape = ".", size = 1)+
  scale_x_continuous(breaks = seq(0 , 1914386, 100000), expand = c(0, 0))+
  scale_y_continuous(limits = c(0, 600))+
  labs(x = "NP genome positions", y = "depth") +
  ggtitle(" raw coverage of UP15 rdnp aligned to NP, map > 0 reads") +
  theme(plot.margin=unit(c(1,1,1,1),"cm"),
        legend.position = "bottom",
        panel.grid.minor = element_line(colour="white", size=0.5),
        plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
        axis.text  = element_text(size=18),
        axis.title = element_text(size = 18, face = "bold")) 

ggsave("C:/Users/marcelo/Dropbox/uptake/task_list_marcelo/correccion/coveragegraphs/UP15_raw_map_section_np_whole.png", width = 80, height = 30, units = "cm")

```

